File: src/main/java/accounting/patterns/EntryType.java
package accounting.patterns;

/**
 * Represents the type of an accounting entry.
 * This enum defines the standard entry types used in the accounting system.
 */
public enum EntryType {
    /**
     * Represents a base usage entry type
     */
    BASE_USAGE("Base Usage"),

    /**
     * Represents a service fee entry type
     */
    SERVICE("Service Fee");

    private final String name;

    /**
     * Creates a new EntryType with the specified display name.
     *
     * @param name the display name for this entry type
     */
    EntryType(String name) {
        this.name = name;
    }

    /**
     * Returns the display name of this entry type.
     *
     * @return the display name
     */
    public String getName() {
        return name;
    }

    /**
     * Returns the EntryType with the specified name, if it exists.
     *
     * @param name the name to look up
     * @return the matching EntryType
     * @throws IllegalArgumentException if no EntryType matches the given name
     */
    public static EntryType fromName(String name) {
        for (EntryType type : values()) {
            if (type.getName().equalsIgnoreCase(name)) {
                return type;
            }
        }
        throw new IllegalArgumentException("No EntryType found for name: " + name);
    }
}

File: src/main/typescript/accounting-pattern/EntryType.ts
/** The kind of entry a posting rule produces. */
export class EntryType {
    static readonly BASE_USAGE = new EntryType("Base Usage");
    static readonly SERVICE = new EntryType("Service Fee");

    private readonly name: string;

    constructor(name: string) {
        this.name = name;
    }

    static entryType(name: string): EntryType {
        return new EntryType(name);
    }

    getName(): string {
        return this.name;
    }

    toString(): string {
        return `EntryType{name='${this.name}'}`;
    }
}
